File: Java/src/deepspace/Weapon.java
/**
 * @brief DEEPSPACE (Java). 
 * Class that represents the arms of a space station to increase the shot power
 * @author Mario Megías Mateo and Shao Jie Hu Chen.
 */

package deepspace;

class Weapon {
    
    private static final float DEFAULTPOWER = 1.0f;
    
    private String name;
    private WeaponType type;
    private int uses;
    
    /**
     * @brief Constructor with parameters
     * @param theName Name associated to weapon. 
     * @param theType Type associated to weapon. 
     * @param theUses Uses associated to weapon. 
     */
    Weapon(String theName, WeaponType theType, int theUses) {
        this.name = theName;
        this.type = theType;
        this.uses = theUses;
    }
    
    /**
     * @brief Copy constructor
     * @param instance Another instance of Weapon. 
     */
    Weapon(Weapon instance) {
        this.name = instance.name;
        this.type = instance.type;
        this.uses = instance.uses;
    }

    WeaponToUI getUIversion() {
        return new WeaponToUI(this);
    }
    
    /**
     * @brief Get method
     * @return Type associated to weapon. 
     */
    public WeaponType getType() {
        return this.type;
    }
    
    /**
     * @brief Get method
     * @return Uses associated to weapon. 
     */
    public int getUses() {
        return this.uses;
    }
    
    /**
     * @brief It indicates the power of the weapon.
     * @return Power associated to weapon. 
     */
    public float power() {
        return this.type.getPower();
    }
    
    /**
     * @brief It decreases the uses of the weapon.
     * @return Weapon power if uses > 0. Otherwise, 1.0. 
     */
    public float useIt() {
        if(uses > 0) {
            uses--;
            return this.power();
        } else {
            return DEFAULTPOWER;
        }
    } 

    public String toString() {
        String data = "[";

        data += "Name:" + this.name + "; ";
        data += "WeaponType:" + this.type + "; ";
        data += "Uses:" + this.uses;

        data += "]";
        
        return data;
    }
}

File: Ruby/Weapon.rb
#encoding:utf-8

# +-------------------------------------------------------------------+ #
# | DEEPSPACE (Ruby)                                                  | #
# | Authors: Shao Jie Hu Chen and Mario Megías Mateo                  | #
# +-------------------------------------------------------------------+ #
# | Class which represents the weapons a space station has to power   | #
# | up their energy at shooting.                                      | #
# +-------------------------------------------------------------------+ #

module Deepspace
    
    require_relative 'WeaponType.rb'
    require_relative 'WeaponToUI.rb'

    class Weapon

        @@DEFAULTPOWER = 1.0

        # Constructor
        def initialize(the_name, the_type, the_uses)
            @name = the_name
            @type = the_type
            @uses = the_uses
        end

        # Copy constructor 
        def self.newCopy(instance)
            new(instance.name, instance.type, instance.uses)
        end

        # Getters
        
        attr_reader :name, :type, :uses

        def power
            @type.power
        end

        # It decreases the uses of the weapon.
        def useIt
            if @uses > 0
                @uses = @uses -1
                power
            else
                @@DEFAULTPOWER
            end
        end

        def getUIversion
            return WeaponToUI.new(self)
        end

        # to_s method
        def to_s
            s = "["
            s += "Name: #{name}"
            s += "; Type: #{type.to_s}"
            s += "; Uses: #{uses}" 
            s += "]"
        end

    end # end of class Weapon

end # end of Deepspace